lib/AST/ASTPrinter.cpp

```diff
@@ -2050,18 +2050,21 @@ void PrintAST::printGenericDeclGenericParams(GenericContext *decl) {
       printGenericSignature(GenericSig, PrintParams | InnermostOnly);
 }
 
-void PrintAST::printGenericDeclGenericRequirements(GenericContext *decl) {
-  if (decl->isGeneric()) {
-    if (auto genericSig = decl->getGenericSignature()) {
-      auto baseGenericSig = decl->getParent()
-          ->getGenericSignatureOfContext();
-      printGenericSignature(genericSig, PrintRequirements,
-                            [baseGenericSig](const Requirement &req) {
-                              if (baseGenericSig)
-                                return !baseGenericSig->isRequirementSatisfied(req);
-                              return true;
-                            });
-    }
+void PrintAST::printDeclGenericRequirements(GenericContext *decl) {
+  // Even if the declaration is not generic, it might carry a contextual
+  // where clause.
+  if (!decl->isGeneric() && !decl->getTrailingWhereClause())
```

Comment:
It might be more robust to just check that the decl's generic signature is different (pointer-wise) than the parent context's, and then do the loop over the unstatisfied requirements. Otherwise, it's possible that the precondition goes out of sync (but unlikely, I guess)

Reply:
Right, the where clause is always around no matter how the generic signature computation went. Thanks!

Reply:
While this *is* safer, perhaps it makes sense to also get rid of completely invalid where clauses on non-generic top-level declarations when we diagnose them, since we won't ever need to inspect them anyways?